pkg/apis/core/types.go

```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+	// the claim by kubelet in case that the ResourceClass got removed in
+	// the meantime. It also helps the resource driver to determine whether
+	// it needs to handle a claim that got marked for deletion.
+	DriverName string
+
+	// Allocation is set by the resource driver once a resource has been
+	// allocated successfully. Nil indicates that the resource is not
+	// allocated.
+	Allocation *AllocationResult
+
+	// ReservedFor indicates which entities are currently allowed to use
+	// the claim. Usually those are Pods, but other objects are also
+	// possible as long as they exist as resources in the apiserver. The
+	// ResourceClaim controller in the kube-controller-manager checks for
+	// stale entries that reference non-existent objects and removes them.
+	//
+	// A scheduler must add a Pod that it is scheduling. This must be done
+	// in an atomic ResourceClaim update because there might be multiple
+	// schedulers working on different Pods that compete for access to the
+	// same ResourceClaim, the ResourceClaim might have been marked for
+	// deletion, or even been deallocated already.
+	//
+	// kubelet will check this before allowing a Pod to run because a a
+	// user might have selected a node manually without reserving resources
+	// or a scheduler might have missed that step, for example because it
+	// doesn't support dynamic resource allocation or the feature was
+	// disabled.
+	//
+	// The maximum size is 32 (= [ResourceClaimReservedForMaxSize]). This
+	// is an artificial limit to prevent a completely unbounded field in
+	// the API.
+	ReservedFor []ResourceClaimUserReference
+
+	// DeallocationRequested gets set by the scheduler when it detects the
+	// situation where pod scheduling cannot proceed because some claim was
+	// allocated for a node that cannot provide some other required
+	// resource.
+	//
+	// The driver then needs to deallocate this claim and the scheduler
+	// will try again.
+	//
+	// While DeallocationRequested is set, no new users may be added to
+	// ReservedFor.
+	DeallocationRequested bool
+}
+
+// ReservedForMaxSize is the maximum number of entries in
+// [ResourceClaimStatus.ReservedFor].
+const ResourceClaimReservedForMaxSize = 32
+
+// AllocationResult contains attributed of an allocated resource.
+type AllocationResult struct {
+	// ResourceHandle contains arbitrary data returned by the driver after a
+	// successful allocation. This is opaque for
+	// Kubernetes. Driver documentation may explain to users how to
+	// interpret this data if needed.
+	//
+	// Resource drivers can use this to store some data directly or
+	// cross-reference some other place where information is stored.
+	// This data is guaranteed to be available when a Pod is about
+	// to run on a node, in contrast to the ResourceClass which
+	// may have been deleted in the meantime.
+	//
+	// The maximum size of this field is 16KiB (= [ResourceHandleMaxSize]).
+	ResourceHandle string
```

Comment:
should this be an unstructured type?

Reply:
You mean `runtime.RawExtension`? That would imply that drivers have to store valid JSON here because the apiserver would validate it. I think an opaque string is better.

Reply:
The advantage is that a patch which adds or removes an allocation only needs to touch one field in (ResourceClaimStatus.Allocation). RBAC permissions for drivers can be limited to just the status or even specific fields (once the apiserver supports that).

Reply:
Is that true? The driver already needs to add a finalizer

Reply:
oh you said "one field", not "one API call".

But changing just one field doesn't sound like a good justification. If it can be done with annotations, it should probably be done with annotations.

Reply:
You are right, I had forgotten about the finalizer. So that permission argument is invalid. But the inverse applies: normal users can edit annotations (although they shouldn't) and thus break the driver. In the status, the handle is protected from user mistakes.

Regarding updates, I meant "one API call and one field": in a single status update call it can set the entire "Allocation" field and doesn't need two calls (one for status, one for annotations).